src/hotspot/cpu/x86/gc/z/zBarrierSetAssembler_x86.cpp

```diff
@@ -186,7 +301,208 @@ void ZBarrierSetAssembler::load_at(MacroAssembler* masm,
+    __ push(rcx);
+    __ movptr(rcx, ExternalAddress((address)&ZPointerLoadShift));
+    __ shlq(rnew_zpointer);
+    __ pop(rcx);
+    __ orq(rnew_zpointer, Address(r15_thread, ZThreadLocalData::store_good_mask_offset()));
+  }
+}
+
+static void store_barrier_buffer_add(MacroAssembler* masm,
+                                     Address ref_addr,
+                                     Register tmp1,
+                                     Label& slow_path) {
+  Address buffer(r15_thread, ZThreadLocalData::store_barrier_buffer_offset());
+
+  __ movptr(tmp1, buffer);
+
+  // Combined pointer bump and check if the buffer is disabled or full
+  __ cmpptr(Address(tmp1, ZStoreBarrierBuffer::current_offset()), 0);
+  __ jcc(Assembler::equal, slow_path);
+
+  Register tmp2 = r15_thread;
+  __ push(tmp2);
+
+  // Bump the pointer
+  __ movq(tmp2, Address(tmp1, ZStoreBarrierBuffer::current_offset()));
+  __ subq(tmp2, sizeof(ZStoreBarrierEntry));
+  __ movq(Address(tmp1, ZStoreBarrierBuffer::current_offset()), tmp2);
+
+  // Compute the buffer entry address
+  __ lea(tmp2, Address(tmp1, tmp2, Address::times_1, ZStoreBarrierBuffer::buffer_offset()));
+
+  // Compute and log the store address
+  __ lea(tmp1, ref_addr);
+  __ movptr(Address(tmp2, in_bytes(ZStoreBarrierEntry::p_offset())), tmp1);
+
+  // Load and log the prev value
+  __ movptr(tmp1, Address(tmp1, 0));
+  __ movptr(Address(tmp2, in_bytes(ZStoreBarrierEntry::prev_offset())), tmp1);
+
+  __ pop(tmp2);
+}
+
+void ZBarrierSetAssembler::store_barrier_medium(MacroAssembler* masm,
+                                                Address ref_addr,
+                                                Register tmp,
+                                                bool is_native,
+                                                bool is_atomic,
+                                                Label& medium_path_continuation,
+                                                Label& slow_path,
+                                                Label& slow_path_continuation) const {
+  assert_different_registers(ref_addr.base(), tmp);
+
+  // The reason to end up in the medium path is that the pre-value was not 'good'.
+
+  if (is_native) {
+    __ jmp(slow_path);
+    __ bind(slow_path_continuation);
+    __ jmp(medium_path_continuation);
+  } else if (is_atomic) {
+    // Atomic accesses can get to the medium fast path because the value was a
+    // raw null value. If it was not null, then there is no doubt we need to take a slow path.
+    __ cmpptr(ref_addr, 0);
+    __ jcc(Assembler::notEqual, slow_path);
+
+    // If we get this far, we know there is a young raw null value in the field.
+    // Try to self-heal null values for atomic accesses
+    __ push(rax);
+    __ push(rbx);
+    __ push(rcx);
+
+    __ lea(rcx, ref_addr);
+    __ xorq(rax, rax);
+    __ movptr(rbx, Address(r15, ZThreadLocalData::store_good_mask_offset()));
+
+    __ lock();
+    __ cmpxchgq(rbx, Address(rcx, 0));
```

Comment:
`ref_addr` is not necessarily materialised here?

Reply:
I think it is, yes. But we want to ensure it's in a register that isn't rbx or rax. So I figured I'd just force materialize it in rcx and call it a day. It might be possible to micro optimize this further and even use the live information we have gathered to eliminate some of the spilling, but I'd like to hold off on that until we integrate. It's again only for atomics, and also happens at most once per field.